### PR TITLE
Fortran task-graph headers/API access and testcase for fib

### PR DESCRIPTION
**Pull request by LonelyCat124**
I have this working (I think) for the fib testcase - I can add more of the testcases tomorrow.

I need to update pyotter and check the outputs also look reasonable, but doesn't seem to have been too hard to add.

### CODE REVIEW

File: src/otter/examples/task-graph/omp-tg-fib-f90.F90 [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    ! Tag: wrap a task
+    root = fortran_otterTaskBegin(c_null_ptr)
+    fibn = fib(n, root)
+    Call fortran_otterTaskEnd(root)
+    !$omp end single
+    !$omp end parallel
+
+    print *, n, fibn
+
+    Call fortran_otterTraceStop()
```
Comment (adamtuft) [inline src/otter/examples/task-graph/omp-tg-fib-f90.F90:68]: This should be `fortran_otterTraceFinalise()` instead

File: include/otter/otter-task-graph.h [hunk 2]
```diff
@@ -35,8 +35,8 @@
  * 
  */
 typedef enum otter_task_sync_t {
-    otter_sync_children,
-    otter_sync_descendants
+    otter_sync_children=1,
```
Comment (adamtuft) [inline include/otter/otter-task-graph.h]: Why do these values need to be explicitly set? Do Fortran enums start from 1? I guess this is to match the values provided by [these lines](https://github.com/Otter-Taskification/otter/blob/1736cb9529a978c281dc47f97f67a8afe27c9a1f/src/otter/events/task-graph/otter-task-graph.F90#L2-L5):

```
    enum, bind(c)
        enumerator otter_sync_children
        enumerator otter_sync_descendants
    end enum
```

File: src/otter/examples/task-graph/omp-tg-fib-f90.F90 (hunk 1 above)
Comment (adamtuft) [inline src/otter/examples/task-graph/omp-tg-fib-f90.F90]: This should be `fortran_otterTraceFinalise()` instead

Comment (adamtuft) [reply src/otter/examples/task-graph/omp-tg-fib-f90.F90]: I pushed a commit to fix this, I hope that doesn't break anything with this PR

**Comment by adamtuft** [comment]
Thanks very much for putting this together, it looks good! I just added a comment about the explicit enum values. I also corrected the call to `fortran_otterTraceFinalise` in src/otter/examples/task-graph/omp-tg-fib-f90.F90

**Comment by LonelyCat124** [comment]
Hmm, I guess I missed pushing that change, thanks for the fix. I'll build some more examples tomorrow.

**Comment by LonelyCat124** [comment]
I added another example, I think probably this is fine to merge if you're ok with it now. Once this is done I'll try to update the PSyclone implementation as such.

**Comment by adamtuft** [comment]
Thanks for the additional example 👍 did you see my comment on the hardcoded enum values?

**Comment by LonelyCat124** [comment]
No I missed that - where is it?

Edit: I think we can also remove those now too, it should work anyway.

**Comment by adamtuft** [comment]
https://github.com/Otter-Taskification/otter/pull/6#discussion_r1165937058

**Comment by adamtuft** [comment]
Do you think it's ok to remove the hardcoded values from the enum? Don't Fortran enumerators start from 0?

**Comment by LonelyCat124** [comment]
Weird, the link to your comment still isn't working for me. To answer your question:

So I had originally hardcoded them when I couldn't get Fortran enumerators to work.

Fortran enumerators are a bit weird - they aren't a concept in Fortran outside of sort of existing as an "C interoperability" thing, the normal approach would just be to have some value set as parameters in a module.
I think we'd be safe to remove those explicit values but we'd need to check the output from Otter probably with the Fortran examples to be sure (does Otter itself check the value input from `otterTaskSynchroniseTasks`?).

I also noticed the "taskloop" example doesn't have a synchronise task call - and I'm not sure whether I should add one or not, its sort of handled implicitly by the `taskloop`/`single` boundary.

**Comment by adamtuft** [comment]
There is a debug logging statement which prints the value passed to `otterSynchroniseTasks`. Otter doesn't explicitly check the synchronisation mode passed to `otterSynchroniseTasks` but it does check whether the value passed is `otter_sync_descendants`. If it's not, it implicitly assumes that you mean to synchronise child tasks.

If we have a Fortran example which uses both `otter_sync_children` and `otter_sync_descendants` we can use the debug logging in Otter to verify what underlying values the enumerator is passing.

EDIT:

re. synchronisation in the taskloop example - it is necessary to call `otterSynchroniseTasks` to tell Otter that the tasks are synchronised as the OMP runtime synchronises all descendant tasks in the taskloop due to the implicit taskgroup. When you annotate an OMP code directly in this way, Otter doesn't know about events which take place in the OMP runtime e.g. task synchronisation. Otter only gets these events automatically when used via the OMPT plugin.

**Comment by LonelyCat124** [comment]
Ok, I'll try to make a dummy case (probably not a real case) that uses both enumerators to just test for now when I can - fighting to build things on cosma still.

**Comment by adamtuft** [comment]
Just added an edit about the taskloop synchronisation

**Comment by LonelyCat124** [comment]
Does otter have a rule on when the synchronisation call should happen - is it before or after the synchronisation call?

**Comment by adamtuft** [comment]
I think it shouldn't matter, as long as the task context you pass to `otterSynchroniseTasks` is still active.

**Comment by LonelyCat124** [comment]
I removed the explicit values and made a very basic example which seems to output the expected value for synchronise tasks:
```
[i] [otterTraceInitialise            ] Otter environment variables:
[i] [otterTraceInitialise            ] host                           glados.dl.ac.uk
[i] [otterTraceInitialise            ] OTTER_TRACE_PATH               trace
[i] [otterTraceInitialise            ] OTTER_TRACE_NAME               otter_trace
[i] [otterTraceInitialise            ] OTTER_APPEND_HOSTNAME          No
Trace output path:             trace/otter_trace.1206457
[d] [otterTaskContext_alloc          ] allocate task context
[d] [otterTaskContext_init           ] initialised task context: 0
[d] [otterTaskBegin                  ] [0] begin task (child of 18446744073709551615)
[d] [otterSynchroniseTasks           ] synchronise tasks: 0
[d] [otterTaskEnd                    ] [0] end task
[d] [otterTaskContext_delete         ] delete task context: 0
[d] [otterTaskContext_alloc          ] allocate task context
[d] [otterTaskContext_init           ] initialised task context: 1
[d] [otterTaskBegin                  ] [1] begin task (child of 18446744073709551615)
[d] [otterSynchroniseTasks           ] synchronise tasks: 1
[d] [otterTaskEnd                    ] [1] end task
[d] [otterTaskContext_delete         ] delete task context: 1
[d] [otterTraceFinalise              ] === finalising archive ===
```

I'll push the changes and you can check the examples all look ok to you now.

**Comment by adamtuft** [comment]
Thanks, that all looks good to me :+1:

**Review by adamtuft — APPROVED** [review]

**Pull request merged**